Add applier flag

apelisse opened the pull request:
Add mandatory "applier" to apply workflow  …
And add a corresponding flag in kubectl, even though the value is
defaulted in kubectl with "kubectl".

The flag is required for Apply patch-type, but entirely forbidden for
non-apply patch types.

**What type of PR is this?**
/kind feature
/sig api-machinery
/priority important-soon

/assign @jennybuckley @kwiesmueller 

Since we talked about this, I think you should like it.
/cc @smarterclayton @liggitt 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

Commit:
Add mandatory "applyManager" to apply workflow

And add a corresponding flag in kubectl, even though the value is
defaulted in kubectl with "kubectl".

The flag is required for Apply patch-type, but entirely forbidden for
non-apply patch types.

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -525,6 +525,11 @@ type PatchOptions struct {
 	Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+
+	// Applier is the name of the applier. This field is required
+	// for apply requests but forbidden for non-apply patch types.
```
Probably be good to be explicit here about the patch type.

kwiesmueller commented:
Seems like the tests in k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration are failing. Rest looks cool.

Should kubectl really default to kubectl as applier and not apply as we did so far?

apelisse commented:
I forgot that we have an apply test in rbac.

smarterclayton commented on pkg/kubectl/cmd/util/helpers.go:
```diff
@@ -402,6 +402,7 @@ func AddDryRunFlag(cmd *cobra.Command) {
 func AddServerSideApplyFlags(cmd *cobra.Command) {
 	cmd.Flags().Bool("server-side", false, "If true, apply runs in the server instead of the client. This is an alpha feature and flag.")
 	cmd.Flags().Bool("force-conflicts", false, "If true, server-side apply will force the changes against conflicts. This is an alpha feature and flag.")
+	cmd.Flags().String("applier", "kubectl", "Name of the applier used to track field ownership. This is an alpha feature and flag.")
```
For alpha flags we add `experimental-` as a prefix.

apelisse replied:
The three flags here are experimental, should I rename them all?

smarterclayton commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go:
```diff
@@ -100,8 +100,17 @@ func ValidateUpdateOptions(options *metav1.UpdateOptions) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if patchType != types.ApplyPatchType && options.Force != nil {
-		allErrs = append(allErrs, field.Forbidden(field.NewPath("force"), "may not be specified for non-apply patch"))
+	if patchType != types.ApplyPatchType {
```
Needs some unit testing here.

smarterclayton commented on test/integration/apiserver/apply/apply_test.go:
```diff
@@ -132,6 +133,7 @@ func TestCreateOnApplyFailsWithUID(t *testing.T) {
 		Namespace("default").
 		Resource("pods").
 		Name("test-pod-uid").
+		Param("applier", "apply").
```
Seeing how this is used, `applier` seems under-specified and possibly too terse.  As a top level flag on the API, we might want to scope this more such as `applyOwner` or something that stands on its own.

smarterclayton commented on test/integration/auth/rbac_test.go:
```diff
@@ -574,6 +574,9 @@ func TestRBAC(t *testing.T) {
 			if r.verb == "PATCH" {
 				// For patch operations, use the apply content type
 				req.Header.Add("Content-Type", string(types.ApplyPatchType))
+				q := req.URL.Query()
```
Would like to see an integration test that verifies the flag is rejected when not using apply patch and accepted when it is.

apelisse commented:
- Renamed `Applier` to `ApplyManager` across the PR,
- Added unit-test for patch validation code,
- Added integration test for query param required/forbidden,
- **NOT** updated the experimental flag in kubectl

PTAL @smarterclayton

apelisse commented:
Looks like I had to rebase, done.

apelisse commented:
Actually, I want to do it differently. Will re-submit in a different PR.
